src/lapack/backends/mkl_common/mkl_lapack.cxx

```diff
@@ -0,0 +1,2852 @@
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrf_batch_scratchpad_size<std::complex<float>>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *lda,
+    std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrf_batch_scratchpad_size<std::complex<float>>(
+        queue, uplo, n, lda, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrf_batch_scratchpad_size<std::complex<double>>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *lda,
+    std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrf_batch_scratchpad_size<std::complex<double>>(
+        queue, uplo, n, lda, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrs_batch_scratchpad_size<float>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *nrhs,
+    std::int64_t *lda, std::int64_t *ldb, std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrs_batch_scratchpad_size<float>(queue, uplo, n, nrhs, lda, ldb,
+                                                                     group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrs_batch_scratchpad_size<double>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *nrhs,
+    std::int64_t *lda, std::int64_t *ldb, std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrs_batch_scratchpad_size<double>(
+        queue, uplo, n, nrhs, lda, ldb, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrs_batch_scratchpad_size<std::complex<float>>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *nrhs,
+    std::int64_t *lda, std::int64_t *ldb, std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrs_batch_scratchpad_size<std::complex<float>>(
+        queue, uplo, n, nrhs, lda, ldb, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t potrs_batch_scratchpad_size<std::complex<double>>(
+    sycl::queue &queue, oneapi::mkl::uplo *uplo, std::int64_t *n, std::int64_t *nrhs,
+    std::int64_t *lda, std::int64_t *ldb, std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::potrs_batch_scratchpad_size<std::complex<double>>(
+        queue, uplo, n, nrhs, lda, ldb, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t ungqr_batch_scratchpad_size<std::complex<float>>(
+    sycl::queue &queue, std::int64_t *m, std::int64_t *n, std::int64_t *k, std::int64_t *lda,
+    std::int64_t group_count, std::int64_t *group_sizes) {
+    return ::oneapi::mkl::lapack::ungqr_batch_scratchpad_size<std::complex<float>>(
+        queue, m, n, k, lda, group_count, group_sizes);
+}
+template <>
+ONEMKL_EXPORT std::int64_t ungqr_batch_scratchpad_size<std::complex<double>>(
```

Comment:
Looks like very good reuse of code! But I guess it will cost us two device checks: one in these interfaces, and one more inside Intel product. Do you think it doesn't make sense to use C API in case of cpu to skip additional device check for mklcpu?

Reply:
This could work for the computational routines, but probably not the scratchpad routines. LAPACK introduced scratchpad_size() functions with oneMKL and did not create corresponding C APIs (where the call to the routine with lwork=-1 returns the scratchpad size).